Content.Shared/Clothing/EntitySystems/StealthClothingSystem.cs

```diff
@@ -0,0 +1,136 @@
+        if (MetaData(user).EntityLifeStage >= EntityLifeStage.Terminating)
+            return false;
+
+        comp.Enabled = enabled;
+        Dirty(comp);
+
+        var stealth = EnsureComp<StealthComponent>(user);
+        // slightly visible, but doesn't change when moving so it's ok
+        var visibility = enabled ? stealth.MinVisibility + comp.Visibility : stealth.MaxVisibility;
+        _stealth.SetVisibility(user, visibility, stealth);
+        _stealth.SetEnabled(user, enabled, stealth);
+        return true;
+    }
+
+    /// <summary>
+    /// Raise <see cref="AddStealthActionEvent"/> then add the toggle action if it was not cancelled.
+    /// </summary>
+    private void OnGetItemActions(EntityUid uid, StealthClothingComponent comp, GetItemActionsEvent args)
+    {
+        var ev = new AddStealthActionEvent(args.User);
+        RaiseLocalEvent(uid, ev);
+        if (ev.Cancelled)
+            return;
+
+        args.Actions.Add(comp.ToggleAction);
+    }
+
+    /// <summary>
+    /// Raises <see cref="EnableStealthEvent"/> if enabling.
+    /// </summary>
+    private void OnToggleStealth(EntityUid uid, StealthClothingComponent comp, ToggleStealthEvent args)
+    {
+        args.Handled = true;
+        var user = args.Performer;
+        if (comp.Enabled)
+        {
+            SetEnabled(uid, user, false, comp);
+            return;
+        }
+
+        var ev = new EnableStealthEvent(user);
+        RaiseLocalEvent(uid, ev);
+        if (ev.Cancelled)
+            return;
+
+        SetEnabled(uid, user, true, comp);
+    }
+
+    /// <summary>
+    /// Calls <see cref="SetEnabled"/> when server sends new state.
+    /// </summary>
+    private void OnHandleState(EntityUid uid, StealthClothingComponent comp, ref AfterAutoHandleStateEvent args)
+    {
+        // SetEnabled checks if it is the same, so change it to before state was received from the server
+        var enabled = comp.Enabled;
+        comp.Enabled = !enabled;
+        var user = Transform(uid).ParentUid;
+        SetEnabled(uid, user, enabled, comp);
+    }
+
+    /// <summary>
+    /// Force unstealths the user, doesnt remove StealthComponent since other things might use it
+    /// </summary>
+    private void OnUnequipped(EntityUid uid, StealthClothingComponent comp, GotUnequippedEvent args)
+    {
+        SetEnabled(uid, args.Equipee, false, comp);
+    }
+}
+
+/// <summary>
+/// Raised on the stealth clothing when attempting to add an action.
+/// </summary>
+public class AddStealthActionEvent : CancellableEntityEventArgs
+{
+    /// <summary>
+    /// User that equipped the stealth clothing.
+    /// </summary>
+    public EntityUid User;
+
+    public AddStealthActionEvent(EntityUid user)
+    {
+        User = user;
+    }
+}
+
+/// <summary>
+/// Raised on the stealth clothing when the user is attemping to enable it.
+/// </summary>
+public class EnableStealthEvent : CancellableEntityEventArgs
```

Comment:
Call it like "AttemptStealthEvent" or something as enable doesn't sound like it's something that's cancellable.